### PR TITLE
Test bridge pool wasm

### PR DESCRIPTION
**Pull request by batconjurer**
This PR depends on #729 and
* fixes the wasm for submitting a transfer to the ethereum bridge pool. 
* Adds some functionality to a set of testing tools that allow us to test that vps accept changes generated by wasm. 
* Fixes some inconsistencies between erc20 multitoken keys and other multitoken keys (the entirety of which might be addressed later).
* Adds tests that the wasm changes are accepted by the bridge pool vp.

### CODE REVIEW

File: tests/src/native_vp/eth_bridge_pool.rs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    use borsh::{BorshDeserialize, BorshSerialize};
+    use namada::ledger::eth_bridge::bridge_pool_vp::BridgePoolVp;
+    use namada::ledger::eth_bridge::parameters::{
+        Contracts, EthereumBridgeConfig, UpgradeableContract,
+    };
+    use namada::ledger::eth_bridge::storage::bridge_pool::BRIDGE_POOL_ADDRESS;
+    use namada::ledger::eth_bridge::storage::wrapped_erc20s;
+    use namada::ledger::eth_bridge::ADDRESS;
+    use namada::proto::Tx;
+    use namada::types::address::{wnam, xan};
```
Comment (james-chf) [inline tests/src/native_vp/eth_bridge_pool.rs:14]:
```suggestion
    use namada::types::address::{wnam, nam};
```

File: wasm/wasm_source/src/tx_bridge_pool.rs [hunk 2]
```diff
@@ -1,25 +1,61 @@
...
+    let transfer =
+        PendingTransfer::try_from_slice(&signed.data.unwrap()[..]).unwrap();
     // pay the gas fees
-    let GasFee {
+    let GasFee { amount, ref payer } = transfer.gas_fee;
+    token::transfer(
+        ctx,
+        payer,
+        &bridge_pool::BRIDGE_POOL_ADDRESS,
+        &address::xan(),
```
Comment (james-chf) [inline wasm/wasm_source/src/tx_bridge_pool.rs:19]:
```suggestion
        &address::nam(),
```

File: tests/src/vm_host_env/tx.rs [hunk 3]
```diff
@@ -171,6 +180,22 @@ impl TestTxEnv {
...
+            &self.storage,
+            &mut self.write_log,
+            &mut self.gas_meter,
+            &self.tx.code,
+            self.tx.data.as_ref().unwrap_or(&empty_data),
+            &mut self.vp_wasm_cache,
+            &mut self.tx_wasm_cache,
+        )
+        .and(Ok(()))
+    }
```
Comment (james-chf) [inline tests/src/vm_host_env/tx.rs:198]: 🎊 worth it's own PR back to `main`!

File: shared/src/ledger/eth_bridge/storage/wrapped_erc20s.rs [hunk 4]
```diff
@@ -35,7 +35,7 @@ impl Keys {
         self.prefix
             .push(&BALANCE_KEY_SEGMENT.to_owned())
             .expect("should always be able to construct this key")
-            .push(&format!("#{}", owner.encode()))
+            .push(&owner.to_db_key())
```
Comment (james-chf) [inline shared/src/ledger/eth_bridge/storage/wrapped_erc20s.rs:38]: yeah this way is much better, the previous `format` way was a hack

**Review by james-chf — CHANGES_REQUESTED** [review]
some `xan`s need to be come `nam`s

File: wasm/wasm_source/src/tx_bridge_pool.rs [hunk 5]
```diff
@@ -1,25 +1,61 @@
...
+        ref asset,
+        ref sender,
         amount,
-        ref payer,
-    } = transfer.gas_fees;
-    token::transfer(payer, &BRIDGE_POOL_ADDRESS, &address::xan(), amount);
+        ..
+    } = transfer.transfer;
+    // if minting wNam, escrow the correct amount
+    if *asset == native_erc20_address(ctx) {
```
Comment (sug0) [inline wasm/wasm_source/src/tx_bridge_pool.rs:30]: nit: does this `asset` need a `ref` here? I don't see it being used anywhere else, so maybe you could just do the comparison: `asset == native_erc20_address(ctx)`

File: shared/src/types/ethereum_events.rs [hunk 6]
```diff
@@ -108,6 +111,24 @@ impl From<EthAddress> for String {
     }
 }
 
+impl KeySeg for EthAddress {
+    fn parse(string: String) -> crate::types::storage::Result<Self>
+    where
+        Self: Sized,
```
[the remark this answers is not in the document]
Comment (sug0) [reply shared/src/types/ethereum_events.rs:117]: I think in traits with methods returning `Self`, we need to constrain the implementing type to have a non-zero size.

by removing the constraint from the trait decl, we get this:
```
error[E0277]: the size for values of type `Self` cannot be known at compilation time
   --> shared/src/types/storage.rs:591:33
    |
591 |     fn parse(string: String) -> Result<Self>;
    |                                 ^^^^^^^^^^^^ doesn't have a size known at compile-time
    |
```

File: tests/src/native_vp/eth_bridge_pool.rs [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+        // initialize Ethereum bridge storage
+        config.init_storage(&mut env.storage);
+        // initialize Bertha's account
+        env.spawn_accounts([&albert_address(), &bertha_address(), &xan()]);
+        // enrich Albert
+        env.credit_tokens(
+            &albert_address(),
+            &xan(),
+            None,
+            BERTHA_WEALTH.into(),
```
[the remark this answers is not in the document]
Comment (batconjurer) [reply tests/src/native_vp/eth_bridge_pool.rs:89]: That bastard!

File: shared/src/types/ethereum_events.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (batconjurer) [reply shared/src/types/ethereum_events.rs:117]: That was IntelliJ generated code. I removed it.

File: tests/src/vm_host_env/tx.rs (hunk 3 above)
Comment (batconjurer) [reply tests/src/vm_host_env/tx.rs:198]: https://github.com/anoma/namada/pull/738

**Review by sug0 — APPROVED** [review]

File: tests/src/vm_host_env/tx.rs [hunk 8]
```diff
@@ -152,9 +153,17 @@ impl TestTxEnv {
...
     ) {
-        let storage_key = token::balance_key(token, target);
+        let storage_key = match &sub_prefix {
+            Some(sub_prefix) => {
+                let prefix =
+                    token::multitoken_balance_prefix(token, sub_prefix);
+                token::multitoken_balance_key(&prefix, target)
+            }
+            None => token::balance_key(token, target),
+        };
```
Comment (james-chf) [inline tests/src/vm_host_env/tx.rs:166]: This change to `TestTxEnv::credit_tokens` should be PRed back to main

File: tests/src/e2e/ledger_tests.rs [hunk 9]
```diff
@@ -109,11 +109,11 @@ fn test_node_connectivity() -> Result<()> {
         NAM,
         "--amount",
         "10.1",
-        "--fee-amount",
+        "--gas-amount",
         "0",
         "--gas-limit",
         "0",
-        "--fee-token",
```
[the remark this answers is not in the document]
Comment (james-chf) [reply tests/src/e2e/ledger_tests.rs:116]: I'll update https://github.com/anoma/namada/pull/764 to not try to remove these flags

**Review by james-chf — CHANGES_REQUESTED** [review]
lgtm, there are a few things to PR back to main:
- `TestTxEnv::execute_tx` fn
- the `namadac tx` flag changes
- `TestTxEnv::credit_tokens` signature change and upates in `wasm/wasm_source`

File: shared/src/ledger/eth_bridge/storage/wrapped_erc20s.rs [hunk 10]
```diff
@@ -58,6 +58,13 @@ impl From<&EthAddress> for Keys {
     }
 }
 
+/// Construct a sub-prefix from an ERC20 address.
+pub fn sub_prefix(address: &EthAddress) -> storage::Key {
+    storage::Key::from(MULTITOKEN_KEY_SEGMENT.to_owned().to_db_key())
+        .push(&address.to_db_key())
+        .expect("should always be able to construct this key")
+}
```
[the remark this answers is not in the document]
Comment (james-chf) [reply shared/src/ledger/eth_bridge/storage/wrapped_erc20s.rs:66]: The function itself, e.g. something like
```rust
    #[test]
    fn test_sub_prefix() {
        let key = sub_prefix(&DAI_ERC20_ETH_ADDRESS);
        assert_matches!(
            &key.segments[..],
            [
                DbKeySeg::StringSeg(multitoken_path),
                DbKeySeg::StringSeg(token_id),
            ] if multitoken_path == MULTITOKEN_KEY_SEGMENT &&
            token_id == &DAI_ERC20_ETH_ADDRESS.to_canonical()
        )
    }
```

**Comment by batconjurer** [comment]
> lgtm, there are a few things to PR back to main:
> 
>     * `TestTxEnv::execute_tx` fn
> 
>     * the `namadac tx` flag changes
> 
>     * `TestTxEnv::credit_tokens` signature change and usages in `wasm/wasm_source`

Covered by PRs #738 and #775

**Comment by batconjurer** [comment]
pls update wasm

**Review by james-chf — APPROVED** [review]

**Pull request merged**